### PR TITLE
add dnsperf wrapper

### PR DESCRIPTION
**Pull request by mfleader**
### Description
This tool's name is `dnsperf`, and it replaces `dns_perf`. It adds a dnsperf benchmark wrapper that stress tests a DNS service, using [dnsperf](https://www.dns-oarc.net/tools/dnsperf). For each load in the stress test, every DNS query's round trip time in microseconds is recorded (along with other relevant information). 

New dependencies in `setup.cfg`, [pydantic](https://pypi.org/project/pydantic/), [python-dateutil](https://pypi.org/project/python-dateutil/), and [toolz](https://pypi.org/project/toolz/), facilitate data model parsing and timestamp string parsing, and functional processing of lists and dictionaries, respectively.

### CODE REVIEW

File: snafu/benchmarks/dnsperf/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+COPY snafu/image_resources/centos8-appstream.repo /etc/yum.repos.d/centos8-appstream.repo
+
+# dnsperf dependencies
+RUN dnf install -y --nodocs epel-release && \
+    dnf install -y --nodocs \
+        ck-devel \
+        openssl-devel \
+        yum-plugin-copr && \
+    dnf copr enable @dnsoarc/dnsperf -y && \
+    dnf install dnsperf -y && \
```
Comment (sjug) [inline snafu/benchmarks/dnsperf/Dockerfile:12]: Be consistent with the flags (like `-y`) , also realistically `-y` is a top level flag:
`dnf -y install package`

Comment (sjug) [inline snafu/benchmarks/dnsperf/Dockerfile:12]: Be consistent with your commands, also `-y` is a top level `dnf` flag:
`dnf -y install package`

Comment (sjug) [reply snafu/benchmarks/dnsperf/Dockerfile:12]: I'd make one mega `dnf install` rather than breaking them out.

**Review by sjug — COMMENTED** [review]
Took a look, only very minor nits in the `Dockerfile`.

In the meantime as I haven't tested it, can you post what a sample output that we'd see upon test completion?

**Comment by jtaleric** [comment]
regarding the document, i recommend you do the following

Basically, data is not the list of results from the test, but a single iteration from the result. so, the uuid for this run would have 56 documents.

```2021-07-28T16:04:36Z - DEBUG - MainProcess - run_snafu: {
    "_index": "snafu-dnsperf-results",
    "_op_type": "create",
    "_source": {
        "address": "127.0.0.1",
        "port": 5454,
        "query_filepath": "/home/mleader/queries.txt",
        "start_time": "2021-07-28 16:04:36",
        "sum_rtt": 0.03285,
        "dnsperf_version": "2.6.0",
        "clients": 1,
        "time_limit": 0.01,
        "transport_mode": "udp",
        "max_allowed_load": 6000.0,
        "cache_size": null,
        "networkpolicy": null,
        "pod_id": null,
        "data": {
            "fqdn": "pi3.lan",
            "rtt_mu_s": 776, 
            "qtype" : "A", 
            "rcode" : 0
         }
        "avg_request_packet_size": 24,
        "avg_response_packet_size": 35,
        "queries_sent": 56,
        "queries_completed": 56,
        "throughput": 1704.718417,
        "workload": "dnsperf",
        "run_id": "NA"
    },
    "_id": "ba34be69f047dcf2bd1748873bc572785195cac21b108ee44356ecfbc4faecdf",
    "run_id": "NA"
}```

File: snafu/benchmarks/dnsperf/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# flake8: noqa
+"""Uperf benchmark"""
```
Comment (learnitall) [inline snafu/benchmarks/dnsperf/__init__.py:4]: Can this be changed to Dnsperf?

File: snafu/benchmarks/dnsperf/dnsperf.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                return None
+            elif sample.successful.stdout is None:
+                self.logger.critical(
+                    (
+                        "dnsperf ran successfully, but did not get output on stdout.\n"
+                        f"Got results: {sample}\n"
+                    )
+                )
+                return None
+
```
Comment (learnitall) [inline snafu/benchmarks/dnsperf/dnsperf.py:197]: Can we add a log message here saying that dnsperf did run successfully? Maybe at an info level?

**Review by learnitall — APPROVED** [review]
This looks great! Left a couple of tiny comments, but other than that LGTM.

File: snafu/benchmarks/dnsperf/dnsperf.py (hunk 3 above)
Comment (mfleader) [reply snafu/benchmarks/dnsperf/dnsperf.py:216]: yep

File: snafu/benchmarks/dnsperf/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,17 @@
+FROM registry.centos.org/centos:8
+
+RUN dnf -y install --nodocs \
```
Comment (jtaleric) [inline snafu/benchmarks/dnsperf/Dockerfile:3]: Please break out each DNF command into its on RUN command.

**Comment by learnitall** [comment]
Can you add the `dnsperf-template.xml` file into the `MANIFEST.in` at some point? Otherwise won't be included in distributions or installations that don't use `pip install -e`

**Comment by jtaleric** [comment]
Given the age @mfleader can we close this PR?

**Comment by mfleader** [comment]
yeah, we can close it

**Pull request closed without merging**